libraries/Adafruit_TinyUSB_Arduino/src/arduino/ports/esp32/Adafruit_TinyUSB_esp32.cpp

```diff
@@ -0,0 +1,166 @@
+// [-fpermissive] #include "hal/gpio_ll.h"
+
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+
+#include "driver/gpio.h"
+#include "driver/periph_ctrl.h"
+
+#include "esp_efuse.h"
+#include "esp_efuse_table.h"
+#include "esp_rom_gpio.h"
+
+#include "esp32-hal.h"
+
+#include "esp32s2/rom/usb/chip_usb_dw_wrapper.h"
+#include "esp32s2/rom/usb/usb_dc.h"
+#include "esp32s2/rom/usb/usb_persist.h"
+
+#include "arduino/Adafruit_TinyUSB_API.h"
+#include "arduino/Adafruit_USBD_Device.h"
+
+//--------------------------------------------------------------------+
+// MACRO TYPEDEF CONSTANT ENUM DECLARATION
+//--------------------------------------------------------------------+
+
+static void configure_pins(usb_hal_context_t *usb) {
+  for (const usb_iopin_dsc_t *iopin = usb_periph_iopins; iopin->pin != -1;
+       ++iopin) {
+    if ((usb->use_external_phy) || (iopin->ext_phy_only == 0)) {
+      esp_rom_gpio_pad_select_gpio(iopin->pin);
+      if (iopin->is_output) {
+        esp_rom_gpio_connect_out_signal(iopin->pin, iopin->func, false, false);
+      } else {
+        esp_rom_gpio_connect_in_signal(iopin->pin, iopin->func, false);
+        if ((iopin->pin != GPIO_FUNC_IN_LOW) &&
+            (iopin->pin != GPIO_FUNC_IN_HIGH)) {
+          // workaround for compiler error with including "hal/gpio_ll.h"
+          // gpio_ll_input_enable(&GPIO, (gpio_num_t) iopin->pin);
+          PIN_INPUT_ENABLE(GPIO_PIN_MUX_REG[iopin->pin]);
+        }
+      }
+      esp_rom_gpio_pad_unhold(iopin->pin);
+    }
+  }
+  if (!usb->use_external_phy) {
+    gpio_set_drive_capability((gpio_num_t)USBPHY_DM_NUM, GPIO_DRIVE_CAP_3);
+    gpio_set_drive_capability((gpio_num_t)USBPHY_DP_NUM, GPIO_DRIVE_CAP_3);
+  }
+}
+
+//--------------------------------------------------------------------+
+// Porting API
+//--------------------------------------------------------------------+
+
+#define USBD_STACK_SZ (4096)
+
+// USB Device Driver task
+// This top level thread processes all usb events and invokes callbacks
+static void usb_device_task(void *param) {
+  (void)param;
+  // RTOS forever loop
+  while (1) {
+    tud_task();
+  }
+}
+
+void TinyUSB_Port_InitDevice(uint8_t rhport) {
+  (void)rhport;
+
+  // Reset USB module
+  periph_module_reset(PERIPH_USB_MODULE);
+  periph_module_enable(PERIPH_USB_MODULE);
+
+  usb_hal_context_t hal = {.use_external_phy = false};
+  usb_hal_init(&hal);
+  configure_pins(&hal);
+
+  // reset core, should be in dcd_esp32sx.c (do that later with more proper testing)
+  USB0.grstctl |= USB_CSFTRST;
+  while ((USB0.grstctl & USB_CSFTRST) == USB_CSFTRST) {}
+
+  tusb_init();
+
+  // Create a task for tinyusb device stack
+  xTaskCreate(usb_device_task, "usbd", USBD_STACK_SZ, NULL,
+              configMAX_PRIORITIES - 1, NULL);
+}
+
+void TinyUSB_Port_EnterDFU(void) {
+  // Reset to Bootloader
+
+  // Reset USB Core
+  USB0.grstctl |= USB_CSFTRST;
+  while ((USB0.grstctl & USB_CSFTRST) == USB_CSFTRST) {}
```

Comment:
@me-no-dev I have tried lots of way, reseting USB OTG core is the actual the best way to have a reliable restart to download mode. Otherwise the endpoint/fifo configure etc can be left-over and could cause failed enumeration while mcu is in download mode. I think the usb core should also give this a try as well.